storm/generic/mailbox.c

```diff
@@ -270,113 +261,104 @@ return_type mailbox_send(mailbox_id_type mailbox_id, message_parameter_type *mes
         return STORM_RETURN_MAILBOX_UNAVAILABLE;
     }
 
+    int full_length = sizeof(message_type) + message_parameter->length;
+
     // If the message won't ever fit into this mailbox, fail.
-    if (message_parameter->length + sizeof (message_parameter_type) >
-        mailbox->total_size)
+    if (full_length > mailbox->queue->size)
     {
         mutex_kernel_signal(&tss_tree_mutex);
         DEBUG_MESSAGE(DEBUG,
                       "Message was larger than the mailbox! (%u > %u)",
-                      message_parameter->length + sizeof (message_parameter_type),
-                      mailbox->total_size);
+                      full_length, mailbox->queue->size);
 
         return STORM_RETURN_MAILBOX_MESSAGE_TOO_LARGE;
     }
 
-    if (message_parameter->length + sizeof (message_parameter_type) >
-        mailbox->free_size)
+    // FIXME: This check is silly and incomplete. Publishing the message might still fail, since
+    // maximum_message_size is a constant instead of determined from the mailbox size. This will
+    // only affect mailboxes created with a size different than the default (64 KiB), e.g.
+    // service mailboxes.
```

Comment:
This is a real limitation of the current code. `maximum_message_size` would have to be calculated for a particular mailbox. I hadn't had time & energy to look into; it will have to remain a `FIXME` for now.

As long as we use (mostly) fixed size mailboxes, this is not a huge problem anyway.